File: day11/day11.nim
import std/[sequtils, algorithm]

type opFunc = proc(x: int): int
type testFunc = proc(x: int): int

type
  Monkey = object
    itms: seq[int]
    operation: opFunc
    test: testFunc
    inspectCount: int


#proc part1(data: string): int =
#  var monkeys = @[
#                  Monkey(
#                    itms: @[97, 81, 57, 57, 91, 61],
#                    operation: proc(x: int): int = (x * 7) div 3,
#                    test: proc(x: int): int = (if (x mod 11 == 0): 5 else: 6),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[88, 62, 68, 90],
#                    operation: proc(x: int): int = (x * 17) div 3,
#                    test: proc(x: int): int = (if (x mod 19 == 0): 4 else: 2),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[74, 87],
#                    operation: proc(x: int): int = (x + 2) div 3,
#                    test: proc(x: int): int = (if (x mod 5 == 0): 7 else: 4),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[53, 81, 60, 87, 90, 99, 75],
#                    operation: proc(x: int): int = (x + 1) div 3,
#                    test: proc(x: int): int = (if (x mod 2 == 0): 2 else: 1),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[57],
#                    operation: proc(x: int): int = (x + 6) div 3,
#                    test: proc(x: int): int = (if (x mod 13 == 0): 7 else: 0),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[54, 84, 91, 55, 59, 72, 75, 70],
#                    operation: proc(x: int): int = (x * x) div 3,
#                    test: proc(x: int): int = (if (x mod 7 == 0): 6 else: 3),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[95, 79, 79, 68, 78],
#                    operation: proc(x: int): int = (x + 3) div 3,
#                    test: proc(x: int): int = (if (x mod 3 == 0): 1 else: 3),
#                    inspectCount: 0),
#                  Monkey(
#                    itms: @[61, 97, 67],
#                    operation: proc(x: int): int = (x + 4) div 3,
#                    test: proc(x: int): int = (if (x mod 17 == 0): 0 else: 5),
#                    inspectCount: 0),
#  ]
#  for round in 0..19:
#    for monkey in mitems(monkeys):
#      for itm in mitems(monkey.itms):
#        let worry = monkey.operation(itm)
#        let dest = monkey.test(worry)
#        monkeys[dest].itms.add(worry)
#      monkey.inspectCount += len(monkey.itms)
#      monkey.itms = @[]
#  let inspections = sorted(monkeys.mapIt(it.inspectCount), Descending)
#  return inspections[0] * inspections[1]


proc part2(data: string): int =
  let commonDivisor = 11 * 19 * 5 * 2 * 13 * 7 * 3 * 17;
  var monkeys = @[
                  Monkey(
                    itms: @[97, 81, 57, 57, 91, 61],
                    operation: proc(x: int): int = (x * 7) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 11 == 0): 5 else: 6),
                    inspectCount: 0),
                  Monkey(
                    itms: @[88, 62, 68, 90],
                    operation: proc(x: int): int = (x * 17) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 19 == 0): 4 else: 2),
                    inspectCount: 0),
                  Monkey(
                    itms: @[74, 87],
                    operation: proc(x: int): int = (x + 2) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 5 == 0): 7 else: 4),
                    inspectCount: 0),
                  Monkey(
                    itms: @[53, 81, 60, 87, 90, 99, 75],
                    operation: proc(x: int): int = (x + 1) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 2 == 0): 2 else: 1),
                    inspectCount: 0),
                  Monkey(
                    itms: @[57],
                    operation: proc(x: int): int = (x + 6) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 13 == 0): 7 else: 0),
                    inspectCount: 0),
                  Monkey(
                    itms: @[54, 84, 91, 55, 59, 72, 75, 70],
                    operation: proc(x: int): int = (x * x) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 7 == 0): 6 else: 3),
                    inspectCount: 0),
                  Monkey(
                    itms: @[95, 79, 79, 68, 78],
                    operation: proc(x: int): int = (x + 3) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 3 == 0): 1 else: 3),
                    inspectCount: 0),
                  Monkey(
                    itms: @[61, 97, 67],
                    operation: proc(x: int): int = (x + 4) mod commonDivisor,
                    test: proc(x: int): int = (if (x mod 17 == 0): 0 else: 5),
                    inspectCount: 0),
  ]
  for round in 0..10000:
    for monkey in mitems(monkeys):
      for itm in mitems(monkey.itms):
        let worry = monkey.operation(itm)
        let dest = monkey.test(worry)
        monkeys[dest].itms.add(worry)
      monkey.inspectCount += len(monkey.itms)
      monkey.itms = @[]
  let inspections = sorted(monkeys.mapIt(it.inspectCount), Descending)
  return inspections[0] * inspections[1]


let data = readFile("small-input.txt")
#data = readFile("input.txt")
echo part2(data)

File: day11/day11.py
from rich import print
from typing import List, Callable, Tuple, Generator
from dataclasses import dataclass, field

@dataclass
class Monkey:
    items: List[int] = field(default_factory=list)
    operation: str = ""
    divisor: int = 0
    dest_true: int = 0
    dest_false: int = 0
    handle_count: int = 0

    def inspect_and_throw_items(self) -> Generator[Tuple[int, int], None, None]:
        for item in self.items:
            worry = eval(self.operation, {}, {"old": item})
            self.handle_count += 1
            if worry % self.divisor == 0:
                yield worry, self.dest_true
            else:
                yield worry, self.dest_false

def init_monkeys(filename: str):
    monkeys: List[Monkey] = []
    cur_monkey = Monkey()
    for line in open(filename).readlines():
        match line.strip().split(":"):
            case s if s[0].startswith("Monkey"):
                cur_monkey = Monkey()
                monkeys.append(cur_monkey)
            case ["Starting items", rest]:
                cur_monkey.items.extend([int(x) for x in rest.split(", ")])
            case ["Operation", rest]:
                cur_monkey.operation =rest.split("=")[-1].strip() 
            case ["Test", rest]:
                cur_monkey.divisor = int(rest.split()[-1])
            case ["If true", rest]:
                cur_monkey.dest_true = int(rest.split()[-1])
            case ["If false", rest]:
                cur_monkey.dest_false = int(rest.split()[-1])
    return monkeys


def part1(filename) -> int:
    monkeys: List[Monkey] = init_monkeys(filename)
    print(monkeys)
    for _ in range(10000):
        for monkey in monkeys:
            for (item, dest) in monkey.inspect_and_throw_items():
                monkeys[dest].items.append(item)
            monkey.items = []
    results = sorted(monkeys, key=lambda x: x.handle_count, reverse=True)
    return results[0].handle_count * results[1].handle_count  


def part2(filename) -> int:
    monkeys: List[Monkey] = init_monkeys(filename)
    print(monkeys)
    for _ in range(10000):
        for monkey in monkeys:
            for (item, dest) in monkey.inspect_and_throw_items():
                monkeys[dest].items.append(item)
            monkey.items = []
    results = sorted(monkeys, key=lambda x: x.handle_count, reverse=True)
    return results[0].handle_count * results[1].handle_count  

print(part1("small-input.txt"))
print(part2("small-input.txt"))
